snuba/stateful_consumer/states/bootstrap.py

```diff
@@ -1,25 +1,101 @@
 import logging
+import json
 
-from typing import Sequence, Tuple
+from typing import Optional, Sequence, Tuple
 from confluent_kafka import Consumer, Message, TopicPartition
 
-from snuba import settings
-from snuba.consumers.strict_consumer import CommitDecision, StrictConsumer
 from snuba.stateful_consumer import ConsumerStateData, ConsumerStateCompletionEvent
 from snuba.utils.state_machine import State
-
+from snuba.consumers.strict_consumer import CommitDecision, StrictConsumer
+from snuba.stateful_consumer.control_protocol import (
+    parse_control_message,
+    SnapshotInit,
+    SnapshotAbort,
+    SnapshotLoaded,
+    ControlMessage,
+)
+from snuba import settings
 
 logger = logging.getLogger('snuba.snapshot-load')
 
 
+class RecoveryState:
+    """
+    Contains the logic that decides what to do for every message read on the
+    control topic.
+    It knows which messages should be taken into account and which ones should
+    be discarded. Knowing this, it is able to tell what to commit.
+    """
+
+    def __init__(self):
+        self.__active_snapshot_msg = None
+        self.__processed_snapshots = set()
+        self.__completion_event = ConsumerStateCompletionEvent.NO_SNAPSHOT
+
+    def get_completion_event(self) -> ConsumerStateCompletionEvent:
+        return self.__completion_event
+
+    def get_active_snapshot_msg(self) -> Optional[ControlMessage]:
+        return self.__active_snapshot_msg
+
+    def process_init(self, msg: SnapshotInit) -> None:
+        logger.debug("Processing init message for %r", msg.id)
+        if msg.product != settings.SNAPSHOT_LOAD_PRODUCT:
+            return
+        if self.__active_snapshot_msg:
+            if isinstance(self.__active_snapshot_msg, SnapshotInit):
```

Comment:
When would this not be true? What type is `__active_snapshot_msg`?

Later: OK. I saw this assigned again below. Ideally the design of this class would be explained in the docstring or a comment or something, rather than requiring the reader to read the class first and figure it out for themself from the implementation

Reply:
Ok, I will beef up the docstring